### PR TITLE
Validate vertex entry point returns the position built-in

### PR DESCRIPTION
**Pull request by teoxoy**
fixes https://github.com/gfx-rs/wgpu/issues/3550

### CODE REVIEW

File: src/valid/interface.rs [hunk 1]
```diff
@@ -353,7 +355,6 @@ impl VaryingContext<'_> {
                 .map_err(|e| e.with_span_context(span_context)),
             None => {
                 match self.types[ty].inner {
-                    //TODO: check the member types
```
Comment (jimblandy) [inline src/valid/interface.rs:356]: Thanks for removing this. It's covered in `validate_impl` now, right?

File: src/valid/interface.rs [hunk 2]
```diff
@@ -369,7 +370,6 @@ impl VaryingContext<'_> {
                                     #[cfg(not(feature = "validate"))]
                                     let _ = index;
                                 }
-                                // TODO: shouldn't this be validate?
```
Comment (jimblandy) [inline src/valid/interface.rs:372]: I think this TODO meant, shouldn't this be conditional on the `"validate"` feature. I think we shouldn't be calling this function at all when `"validate"` isn't enabled. But that's not something for this PR; filed as #2302.

File: src/valid/interface.rs [hunk 3]
```diff
@@ -545,6 +545,33 @@ impl super::Validator {
             return Err(EntryPointError::UnexpectedWorkgroupSize.with_span());
         }
 
+        #[cfg(feature = "validate")]
+        if ep.stage == crate::ShaderStage::Vertex {
+            let has_position = match ep.function.result {
```
Comment (jimblandy) [inline src/valid/interface.rs:550]: If you check for this just after calling `VaryingContext::validate` on `ep.function.result`, then `crate::BuiltIn::Position { invariant: false }` will be in `result_built_ins` if it was present anywhere on the entry point's return value. You'll have to arrange to check even when `ep.function.result` is `None`, but even so, checking that might be simpler than doing a scan here.

File: tests/snapshots.rs [hunk 4]
```diff
@@ -117,7 +117,7 @@ fn check_targets(module: &naga::Module, name: &str, targets: Targets) {
 
     let info = naga::valid::Validator::new(naga::valid::ValidationFlags::all(), capabilities)
         .validate(module)
-        .expect("Naga module validation failed");
+        .expect(&format!("Naga module validation failed on test '{name}'"));
```
Comment (jimblandy) [inline tests/snapshots.rs:120]: :heart:

File: tests/out/wgsl/binop-frag.wgsl
Comment (jimblandy) [inline tests/out/wgsl/binop-frag.wgsl:1]: Was this just a random file abandoned in the output directory?

**Review by jimblandy — CHANGES_REQUESTED** [review]
I was wondering if it was right to enforce this on all modules, not just those derived from WGSL. But:

- GLSL says the vertex position is undefined if the vertex shader doesn't assign to `gl_Position`, so it seems like any vertex shader that doesn't do so should probably be flagged by Naga. 

- I can't find anything in the Vulkan or SPIR-V specifications that justifies enforcing this rule on SPIR-V input, but I also don't see the harm.

So, I guess we might as well go with this and see if anyone complains. Thanks for updating all the tests.

I think there's a better way to do the check.

File: src/valid/interface.rs (hunk 2 above)
Comment (teoxoy) [reply src/valid/interface.rs:372]: Ah, I see. I thought it meant using the validate fn.

File: src/valid/interface.rs (hunk 1 above)
Comment (teoxoy) [reply src/valid/interface.rs:356]: Yep

File: tests/out/glsl/multiview.main.Fragment.glsl
[the remark this answers is not in the document]
Comment (teoxoy) [reply tests/out/glsl/multiview.main.Fragment.glsl:1]: This is an odd one, the input test file for this should have been `tests/in/multiview.wgsl` which never wrote to the position.

File: src/valid/interface.rs (hunk 3 above)
Comment (teoxoy) [reply src/valid/interface.rs:550]: Good idea 👍

**Review by jimblandy — APPROVED** [review]
Thanks!

**Pull request merged**